content/program-testing.md

```diff
@@ -0,0 +1,513 @@
+
+Tests are defined in the `tests` module with the `#[test]` attribute. When running `cargo test`, every function inside this module marked as a test will be run. You can also create helper functions that are not tests in the module, just don’t annotate them with the `#[test]` attribute.
+
+### How to build unit tests
+
+To build unit tests in Rust, you will have to use the [`solana_sdk`](https://docs.rs/solana-sdk/latest/solana_sdk/) crate. This crate is essentially the same thing as the `@solana/web3.js` package that we’ve been using in Typescript and gives us a way to interact with Solana programs in Rust. Another crate that will be useful and was made specifically for testing Solana programs is [`solana_program_test`](https://docs.rs/solana-program-test/latest/solana_program_test/#) which contains a BanksClient-based testing framework.
+
+A simple example of a unit test residing inside a `processor.rs` file may look like
+
+```rust
+// Inside processor.rs
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        assert_matches::*,
+        solana_program::instruction::{AccountMeta, Instruction},
+        solana_program_test::*,
+        solana_sdk::{signature::Signer, transaction::Transaction, signer::keypair::Keypair},
+    };
+
+    #[tokio::test]
+    async fn it_works() {
+        let program_id = Pubkey::new_unique();
+
+        let (mut banks_client, payer, recent_blockhash) = ProgramTest::new(
+            "bpf_program_template",
+            program_id,
+            processor!(process_instruction),
+        )
+        .start()
+        .await;
+
+        let test_acct = Keypair::new();
+
+        let mut transaction = Transaction::new_with_payer(
+            &[Instruction {
+                program_id,
+                accounts: vec![
+                    AccountMeta::new(payer.pubkey(), true),
+                    AccountMeta::new(test_acct.pubkey(), true)
+                ],
+                data: vec![1, 2, 3],
+            }],
+            Some(&payer.pubkey()),
+        );
+        transaction.sign(&[&payer, &test_acct], recent_blockhash);
+
+        assert_matches!(banks_client.process_transaction(transaction).await, Ok(_);
+    }
+}
+```
+
+In the code snippet, we created a public key to use as our `program_id` and then initialized a `ProgramTest`. Then, we create a second `Keypair` and built our `Transaction` with the appropriate parameters. Finally, we used the `banks_client` that was returned when calling `ProgramTest::new` to process this transaction and check that the return value is equal to `Ok(_)`. This is a very simple test, but from the code snippet you can see how you would go about creating more complex tests that involve more accounts and data similar to how you would client side.
+
+## Integration tests
+
+## What are integration tests?
```

Comment:
@ixmorrow this seems like an odd choice - isn't the question a sub-section within the "Integration Tests" header?

I recommend changing it back to an h3 `###`